ecs-agent/acs/session/session_test.go

```diff
@@ -0,0 +1,1328 @@
+                    "privateAddress": "ipv4"
+                }],
+                "ipv6Addresses": [{
+                    "address": "ipv6"
+                }],
+                "subnetGatewayIpv4Address": "ipv4/20",
+                "macAddress": "mac"
+        }],
+        "roleCredentials": {
+          "credentialsId": "credsId",
+          "accessKeyId": "accessKeyId",
+          "expiration": "2016-03-25T06:17:19.318+0000",
+          "roleArn": "r1",
+          "secretAccessKey": "secretAccessKey",
+          "sessionToken": "token"
+        },
+        "version": "3",
+        "volumes": [],
+        "family": "f",
+        "arn": "arn",
+        "desiredStatus": "RUNNING"
+      }
+    ],
+    "generatedAt": 1,
+    "clusterArn": "1",
+    "containerInstanceArn": "1",
+    "seqNum": 1
+  }
+}
+`
+	sampleRefreshCredentialsMessage = `
+{
+  "type": "IAMRoleCredentialsMessage",
+  "message": {
+    "messageId": "123",
+    "clusterArn": "default",
+    "taskArn": "t1",
+    "roleType": "TaskApplication",
+    "roleCredentials": {
+      "credentialsId": "credsId",
+      "accessKeyId": "newakid",
+      "expiration": "later",
+      "roleArn": "r1",
+      "secretAccessKey": "newskid",
+      "sessionToken": "newstkn"
+    }
+  }
+}
+`
+	sampleAttachResourceMessage = `
+{
+  "type": "ConfirmAttachmentMessage",
+  "message": {
+    "messageId": "123",
+    "clusterArn": "arn:aws:ecs:us-west-2:123456789012:cluster/a1b2c3d4-5678-90ab-cdef-11111EXAMPLE",
+    "containerInstanceArn": "arn:aws:ecs:us-west-2:123456789012:container-instance/a1b2c3d4-5678-90ab-cdef-11111EXAMPLE",
+    "taskArn": "arn:aws:ecs:us-west-2:1234567890:task/test-cluster/abc",
+    "waitTimeoutMs": 1000,
+    "attachment": {
+      "attachmentArn": "arn:aws:ecs:us-west-2:123456789012:ephemeral-storage/a1b2c3d4-5678-90ab-cdef-11111EXAMPLE",
+      "attachmentProperties": [
+        {
+          "name": "resourceID",
+          "value": "id1"
+        },
+        {
+          "name": "volumeID",
+          "value": "id1"
+        },
+        {
+          "name": "volumeSizeInGiB",
+          "value": "size1"
+        },
+        {
+          "name": "requestedSizeInGiB",
+          "value": "size1"
+        },
+        {
+          "name": "resourceType",
+          "value": "EphemeralStorage"
+        },
+        {
+          "name": "deviceName",
+          "value": "device1"
+        }
+      ]
+    }
+  }
+}
+`
+	agentVersion      = "1.23.4"
+	agentGitShortHash = "ffffffff"
+	dockerVersion     = "1.2.3"
+	acsURL            = "http://endpoint.tld"
+)
+
+var inactiveInstanceError = errors.New("InactiveInstanceException")
+var noopFunc = func() {}
+var testCreds = credentials.NewStaticCredentials("test-id", "test-secret", "test-token")
+var testMinAgentConfig = &wsclient.WSClientMinAgentConfig{
+	AcceptInsecureCert: true,
+	AWSRegion:          "us-west-2",
+	DockerEndpoint:     "unix:///var/run/docker.sock",
+	IsDocker:           true,
+}
```

Comment:
thank you for all the robust testing! should this be moved to testconst or a common_test file to decrease the size here?

Reply:
No problem, thanks for taking a look at the pull request!

Typically for test constants that are used across multiple files, I move those to be in testconst package. However, since these are only used in session testing, I have left them here in the session test file (similar to other component-specific test constants in other test files under `ecs-agent/acs/session` directory).

Reply:
I see your point, it may get somewhat difficult to follow these tests with so much testing setup but it is non-blocking.

As a follow-up, is there a reason you hardcoded an `IAMRoleCredentialsMessage` for `sampleRefreshCredentialsMessage` instead of using the hard type provided by `ecsacs`?

If not, would it be easier to use a real `ecsacs.IAMRoleCredentialsMessage` in case its struct fields change?

Reply:
> As a follow-up, is there a reason you hardcoded an IAMRoleCredentialsMessage for sampleRefreshCredentialsMessage instead of using the hard type provided by ecsacs?
> 
> If not, would it be easier to use a real ecsacs.IAMRoleCredentialsMessage in case its struct fields change?

This was pulled in from the current `agent/acs/handler/acs_handler_test.go` file. Ultimately it is because the channel used by the test fake ACS server to read in messages is a `string` channel, not a `ecsacs.IAMRoleCredentialsMessage` channel. If we were to make it a channel for a specific type of message, then tests such as `TestStartSessionHandlesAttachResourceMessages` or `TestSessionDoesntLeakGoroutines` would break since those tests needs to have the server able to read messages that are not refresh credentials messages.

Reply:
gotcha, thanks for explaining 👍